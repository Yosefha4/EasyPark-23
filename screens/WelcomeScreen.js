import axios from "axios";
import { useContext, useEffect, useState } from "react";

import { Alert, Image, StyleSheet, Text, View, Platform } from "react-native";
import { AuthContext } from "../store/contextAuth";

import Button from "../components/ui/Button";

import welcImg from "../assets/imageWithoutBg.png";
import { useNavigation } from "@react-navigation/native";

import * as Notifications from "expo-notifications";

import { addDoc, collection, getDocs, query, where } from "firebase/firestore";
import { db } from "../config/firebase";
import EmailContext from "../store/emailContext";

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: true,
    shouldSetBadge: false,
  }),
});

function WelcomeScreen() {
  const [currentPushToken, setCurrentPushToken] = useState("");
  const [userUidToken, setUserUidToken] = useState("");
  const [counterValue, setCounterValue] = useState(0);

  //Get email from Context API
  const emailContext = useContext(EmailContext);
  // console.log(emailContext)
  const email = emailContext.email;
  // console.log("email" ,email)

  const navigation = useNavigation();

  const authCtx = useContext(AuthContext);
  const token = authCtx.token;

  // const currEmail = authCtx.token;

  // useEffect(() => {
  //   configurePushNoti();
  // }, []);

  useEffect(() => {
    // Fetch the current available days from the database and update the state
    const fetchUuidToken = async () => {
      try {
        if (!email) {
          // Email is not available yet, return or handle accordingly
          return;
        }
    
        const matchUuidToken = collection(db, "users");
        const q = query(matchUuidToken, where("userEmail", "==", email));
        const querySnapshot = await getDocs(q);
    
        const availableDaysData = querySnapshot.docs.map((doc) => doc.data());
    
        setUserUidToken(availableDaysData);
        console.log(availableDaysData)
        // console.log("The user UID is:", availableDaysData[0]?.userToken);
        // console.log(userUidToken[0].userToken);
        console.log("The user EMAIL is:", email);
      } catch (error) {
        console.log("welcome screen, error", error);
      }
    };

    fetchUuidToken();
  }, [currentPushToken, counterValue]);

  const updateStateRepeatedly = (times) => {
    let counter = 0;
    const intervalId = setInterval(() => {
      setCounterValue((prevCount) => prevCount + 1);
      counter++;

      if (counter === times) {
        clearInterval(intervalId);
      }
    }, 3000);
  };

  useEffect(() => {
    updateStateRepeatedly(4);
  }, []);

  useEffect(() => {
    async function callToConfig() {
      await configurePushNoti();
    }
    callToConfig();
  }, []);

  // console.log("the the the " , userUidToken[0].userToken)

  async function configurePushNoti() {
    const { status } = await Notifications.getPermissionsAsync();
    let finalStatus = status;

    if (finalStatus !== "granted") {
      const { status } = await Notifications.requestPermissionsAsync();
      finalStatus = status;
    }

    if (finalStatus !== "granted") {
      Alert.alert("Permission required!", "push Noti need permission");
      return;
    }
    const pushToken = await Notifications.getExpoPushTokenAsync();
    // console.log(pushToken);
    setCurrentPushToken(pushToken);
    if (Platform.OS === "android") {
      Notifications.setNotificationChannelAsync("default", {
        name: "default",
        importance: Notifications.AndroidImportance.DEFAULT,
      });
    }
  }

  // async function addUserTokenDoc(userEmail, userToken) {
  //   try {
  //     const userNtoken = collection(db, "users");
  //     await addDoc(userNtoken, {
  //       userEmail: userEmail,
  //       userToken: token,
  //     });
  //     Alert.alert("משתמש נוצר בהצלחה!");
  //   } catch (error) {
  //     console.log(error);
  //     Alert.alert("משהו השתבש...", " אנא נסה שוב מאוחר יותר או צור איתנו קשר ");
  //   }
  // }

  async function saveNotificationToken() {
    console.log("userUIDToken : " , userUidToken)
    if(userUidToken.length > 1){
      try {
        const notifications = collection(db, "pushNotif");
  
  
        const querySnapshot = await getDocs(
          query(notifications, where("userIDToken", "==", userUidToken[0].userToken))
        );
  
        if (querySnapshot.empty) {
          const userIDToken = userUidToken && userUidToken[0]?.userToken;
          const pushToken = currentPushToken || "Undefined pushToken";
          // await configurePushNoti();
          await addDoc(notifications, {
            userIDToken: userIDToken || "Undefined userIDToken",
            pushToken: pushToken,
          });
          Alert.alert("ה-Push Token נשמר בהצלחה !");
        } else {
          console.log("The token already exist ...");
        }
      } catch (error) {
        console.log(error);
        Alert.alert("משהו השתבש...", "WELCOMESCREEN אנא נסה שוב מאוחר יותר או צור איתנו קשר ");
      }
    }

   
  }

  // console.log("TOKEN : ", token);

  return (
    <View style={styles.rootContainer}>
      <Text testID="welcTestText" style={styles.title}>
        ברוכים הבאים !
      </Text>

      <View style={styles.image_container}>
        <Image style={styles.image} source={welcImg} testID="welcomeImage" />
      </View>

      <Button
        testID="welcomeButton"
        onPress={() => {
          saveNotificationToken();
          navigation.navigate("Personal", { token });
        }}
      >
        המשך{" "}
      </Button>
    </View>
  );
}

export default WelcomeScreen;

const styles = StyleSheet.create({
  rootContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    padding: 32,
  },
  image_container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
    height: "100%",
  },
  image: {
    position: "absolute",
    width: "100%",
    height: "80%",
    zIndex: -1,
  },
  title: {
    fontSize: 20,
    fontWeight: "bold",
    marginBottom: 8,
  },
});
