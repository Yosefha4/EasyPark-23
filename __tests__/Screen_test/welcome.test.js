import React from "react";
import renderer from "react-test-renderer";
import WelcomeScreen from "../../screens/WelcomeScreen"
import AuthContextProvider, { AuthContext } from "../../store/contextAuth";
import { NavigationContainer, useNavigation } from "@react-navigation/native";
import { View } from "react-native";
import Button from "../../components/ui/Button";



// import {welcImgTest} from '../assets/welcImg'

const mockToken = "1234abcd";

jest.mock("@react-navigation/native", () => ({
  useNavigation: jest.fn(),
}));


jest.mock("expo-notifications", () => ({
  getPermissionsAsync: jest.fn().mockResolvedValue({ status: "granted" }),
  requestPermissionsAsync: jest.fn().mockResolvedValue({ status: "granted" }),
  getExpoPushTokenAsync: jest.fn().mockResolvedValue("mockedPushToken"),
  setNotificationChannelAsync: jest.fn(),
  setNotificationHandler: jest.fn(),
}));

describe("WelcomeScreen", () => {
  beforeEach(() => {
    useNavigation.mockClear();
  });

  it("shows the correct text", async () => {
    useNavigation.mockReturnValueOnce({
      navigate: jest.fn(),
    });

    const tree = renderer.create(<WelcomeScreen />);
    const textComponent = tree.root.findByProps({ testID: "welcTestText" });

    expect(textComponent.props.children).toEqual("ברוכים הבאים !");
  });

  it("displays the correct image", () => {
    const tree = renderer.create(<WelcomeScreen />);
    const imageComponent = tree.root.findByProps({ testID: "welcomeImage" });

    // console.log(imageComponent.props.source); // Log the value of the source prop

    expect(imageComponent.props.source.testUri).toEqual(
      "../../../assets/imageWithoutBg.png"
    );
  });

  it("should render correctly with mocked AuthContext", () => {
    const mockAuthContext = { token: mockToken };
    const mockNavigation = {
      navigate: jest.fn(),
    };
    const tree = renderer.create(
      <AuthContext.Provider value={mockAuthContext}>
        <WelcomeScreen navigation={mockNavigation} />
      </AuthContext.Provider>
    );
    expect(tree.toJSON()).toMatchSnapshot();

    

  });
});

describe("Welcome Screen Navigation", () => {
  let tree;
  const mockNavigation = { navigate: jest.fn() };
  const mockAuthContext = { token: "mockToken" };

  beforeEach(() => {
    useNavigation.mockReturnValue(mockNavigation);

    tree = renderer.create(
      <AuthContext.Provider value={mockAuthContext}>
        <WelcomeScreen />
      </AuthContext.Provider>
    );
  });

  afterEach(() => {
    useNavigation.mockReset();
    mockNavigation.navigate.mockReset();
  });

  it("navigates to PersonalScreen on button press", () => {
    const button = tree.root.findByProps({ testID: "welcomeButton" });
    button.props.onPress();

    expect(mockNavigation.navigate).toHaveBeenCalledWith("Personal", {
      token: "mockToken",
    });
  });
});
