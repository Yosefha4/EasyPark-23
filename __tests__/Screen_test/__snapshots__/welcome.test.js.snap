// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomeScreen should render correctly with mocked AuthContext 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "padding": 32,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 8,
      }
    }
    testID="welcTestText"
  >
    ברוכים הבאים !
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/imageWithoutBg.png",
        }
      }
      style={
        {
          "height": "80%",
          "position": "absolute",
          "width": "100%",
          "zIndex": -1,
        }
      }
      testID="welcomeImage"
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "backgroundColor": "#0059af",
          "borderRadius": 6,
          "elevation": 2,
          "paddingHorizontal": 12,
          "paddingVertical": 6,
          "shadowColor": "black",
          "shadowOffset": {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
        },
        false,
      ]
    }
  >
    <View>
      <Text
        style={
          {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        המשך
         
      </Text>
    </View>
  </View>
</View>
`;
